codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt

```diff
@@ -0,0 +1,114 @@
+import software.amazon.smithy.rust.codegen.server.typescript.smithy.TsServerCargoDependency
+
+/**
+ * Configure the [lib] section of `Cargo.toml`.
+ *
+ * [lib]
+ * name = "$CRATE_NAME"
+ * crate-type = ["cdylib"]
+ */
+class CdylibManifestDecorator : ServerCodegenDecorator {
+    override val name: String = "CdylibDecorator"
+    override val order: Byte = 0
+
+    override fun crateManifestCustomizations(
+        codegenContext: ServerCodegenContext,
+    ): ManifestCustomizations =
+        mapOf(
+            "lib" to mapOf(
+                "crate-type" to listOf("cdylib"),
+            ),
+        )
+}
+
+class NapiBuildRsDecorator : ServerCodegenDecorator {
+    override val name: String = "NapiBuildRsDecorator"
+    override val order: Byte = 0
+    private val napi_build = TsServerCargoDependency.NapiBuild.toType()
+
+    override fun extras(codegenContext: ServerCodegenContext, rustCrate: RustCrate) {
+        rustCrate.withFile("build.rs") {
+            rustTemplate(
+                """
+                fn main() {
+                    #{napi_build}::setup();
+                }
+                """,
+                "napi_build" to napi_build,
+            )
+        }
+    }
+}
+
+class NapiPackageJsonDecorator : ServerCodegenDecorator {
+    override val name: String = "NapiPackageJsonDecorator"
+    override val order: Byte = 0
+
+    override fun extras(codegenContext: ServerCodegenContext, rustCrate: RustCrate) {
+        val name = codegenContext.settings.moduleName.toSnakeCase()
+        val version = codegenContext.settings.moduleVersion
+
+        // TODO(https://github.com/awslabs/smithy-rs/issues/2317): we should probably use a real JSON writer, but I did not want to add
+        //  other external libraries at this stage.
+        rustCrate.withFile("package.json") {
+            val content = """{
+                "name": "@amzn/$name",
+                "version": "$version",
+                "main": "index.js",
+                "types": "index.d.ts",
+                "napi": {
+                    "name": "$name",
+                    "triple": {}
+                },
+                "devDependencies": {
+                    "@napi-rs/cli": ">=2",
+                    "@types/node": ">=18"
+                },
+                "engines": {
+                    "node": ">=18"
+                },
+                "scripts": {
+                    "artifacts": "napi artifacts",
+                    "build": "napi build --platform --release",
+                    "build:debug": "napi build --platform",
+                    "prepublishOnly": "napi prepublish -t npm",
+                    "universal": "napi universal",
+                    "version": "napi version"
+                },
+                "packageManager": "yarn",
+                "dependencies": {
+                    "yarn": ">=1"
+                }
+}"""
+            this.write(content)
```

Comment:
I think you can write your own `writeJson` function with a `@Language` annotation so that you get syntax highlighting on the json, example: https://github.com/awslabs/smithy-rs/blob/c7c82855986673fc950c532d4cbdfa6cee4cc201/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/RustWriter.kt#L146-L151

Reply:
We would leave it like this for now (see comments above).